File: 64.minimum_path_sum.py
"""
Given a m x n grid filled with non-negative numbers, find a path from top left
to bottom right, which minimizes the sum of all numbers along its path.

Note: You can only move either down or right at any point in time.

"""

class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
        dp = [float('inf')] * (n+1)
        dp[1] = 0
        
        for i in range(m):
            for j in range(n):
                dp[j+1] = min(dp[j], dp[j+1]) + grid[i][j]
        
        return dp[-1]
        
        
        
"""
Setting dp[0] = inf allows ignoring the non-existent left neighbor for the
left-most cells.

Setting dp[1] = 0 allows entering the graph from above at the first column

Setting dp[>1] = inf disallows entering the graph from above, after the first column

The proper way to visualize this is to imagine the augmented maze.  It is still tricky.
"""

File: 64.minimum_path_sum.cc
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
      // initialize first row of dp
      int C=grid[0].size();
      vector<int> buf1(C, INT_MAX);
      buf1[0] = 0;
      vector<int> buf2(C);
      vector<int> *pre = &buf1, *cur = &buf2;
      for (auto row: grid) {
        (*cur)[0] = (*pre)[0] + row[0];
        for (int i=1; i!=C; i++) {
          (*cur)[i] = min((*pre)[i], (*cur)[i-1]) + row[i];
        }
        std::swap(cur, pre);
      }
      return (*pre).back();
    }
};

/*

1 3 1
1 5 1
4 2 1


*/



